docs/tutorials/forum-application/3.send-message.md

```diff
@@ -0,0 +1,381 @@
+
+The method takes in a `PrepareProposalRequest` object, which contains information about the current state of the blockchain,
+such as the current height and the last committed block hash. It may also include other relevant information that the
+application needs to generate the proposal.
+
+The `PrepareProposal` method should return a `PrepareProposalResponse` object, which includes the proposed block contents.
+This typically includes the list of transactions (txs) that should be included in the next block.
+
+**Following is the code for the `PrepareProposal` method:**
+
+```go
+// PrepareProposal is used to prepare a proposal for the next block in the blockchain. The application can re-order, remove
+// or add transactions
+func (app *ForumApp) PrepareProposal(_ context.Context, req *abci.PrepareProposalRequest) (*abci.PrepareProposalResponse, error) {
+    fmt.Println("Executing Application PrepareProposal")
+
+	// Get the curse words from the vote extensions
+	voteExtensionCurseWords := app.getWordsFromVe(req.LocalLastCommit.Votes)
+
+	// Prepare req puts the BanTx first, then adds the other transactions
+	// ProcessProposal should verify this
+	proposedTxs := make([][]byte, 0)
+	finalProposal := make([][]byte, 0)
+	bannedUsersString := make(map[string]struct{})
+	for _, tx := range req.Txs {
+		msg, err := model.ParseMessage(tx)
+		if err != nil {
+			continue
+		}
+		// Adding the curse words from vote extensions too
+		if !hasCurseWord(msg.Message, voteExtensionCurseWords) {
+			proposedTxs = append(proposedTxs, tx)
+		} else {
+			// If the message contains curse words then ban the user by
+			// creating a "ban transaction" and adding it to the final proposal
+			banTx := model.BanTx{UserName: msg.Sender}
+			bannedUsersString[msg.Sender] = struct{}{}
+			resultBytes, err := json.Marshal(banTx)
+			if err == nil {
+				finalProposal = append(finalProposal, resultBytes)
+			} else {
+				// the ban transaction will not be included in the final proposal
+				fmt.Println("ban transaction failed to marshal in PrepareProposal")
+			}
+		}
+	}
+
+	// Need to loop again through the proposed Txs to make sure there is none left by a user that was banned
+	// after the tx was accepted
+	for _, tx := range proposedTxs {
+		// there should be no error here as these are just transactions we have checked and added
+		msg, err := model.ParseMessage(tx)
+		if err != nil {
+			fmt.Println("failed to parse message in PrepareProposal")
+		} else {
+			// If the user is banned then include this transaction in the final proposal
+			if _, ok := bannedUsersString[msg.Sender]; !ok {
+				finalProposal = append(finalProposal, tx)
+			}
+		}
+	}
+	return &abci.PrepareProposalResponse{Txs: finalProposal}, nil
+}
+```
+
+**Explanation of code:**
+
+`PrepareProposal` function first retrieves curse words from a `vote extension.` Then, it iterates over the transactions
```

Comment:
> retrieves curse words from a `vote extension.`

This is not accurate. It retrieves the curse words for all vote extensions received at the end of last height.
Actually, the original use case here is that only words that are flagged as curse words by > 1/3 of validators are actually considered curse words. This is done to present malicious validators from introducing "normal" words as curse words.